Game OSD: Allow rename/delete savestates in-game

garbear opened the pull request:
## Description

This PR applies a series of fixes and improvements to the existing OSD, then adds a feature that lets you rename/delete savestates from within the game.

The menu uses the existing context menu dialog, which doesn't really fit the aesthetics of the game thumbnail dialog.

But at least, though fugly, it functions 100% correctly. Maybe we can special-case the context menu dialog to play nicely with the game thumb dialog.

## Motivation and context

As stated [here](https://github.com/xbmc/xbmc/pull/22490#issuecomment-1399924449) by @KOPRajs, this ability is desired, and it's worse UX when missing.

## How has this been tested?

Testing is currently being done on my latest round of RetroPlayer builds:

Build can be found at https://github.com/garbear/xbmc/releases

## What is the effect on users?

* Allow renaming/deleting savestates from within a game

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

garbear committed:
Games: Remove unused function

garbear committed:
Port dialog: Improve error handling for invalid item index

garbear committed:
Game OSD: Improve error handling for invalid item index

garbear committed:
Game OSD: Allow renaming/deleting savestates in-game

Additional modifications were necessary to improve the behavior of the
in-game savestate dialog when another dialog is opened on top of it.

garbear committed:
Game OSD: Fix conflicting IDs when other dialogs are visible

The game OSD uses window IDs in the range 108XX, so prefix game-osd-specific
control IDs with "108".

This is needed when the YesNo Dialog is shown over the game OSD, because
the YesNo dialog's "OK" button ID is 11. When the "OK" button is focused,
our game thumbnail is erroneously focused too.

garbear committed:
Game OSD: Improvements to out-of-game savestate dialog

This change brings the following improvements:

  * Prepopulate keyboard dialog with savestate name
  * Allow empty savestate label
  * Error dialog when savestate renaming or deletion fails
  * General refactoring

lrusak commented on xbmc/games/dialogs/osd/DialogGameVideoSelect.cpp:
```diff
@@ -29,8 +29,8 @@ using namespace KODI;
 #define CONTROL_HEADING 1
-#define CONTROL_THUMBS 11
-#define CONTROL_DESCRIPTION 12
+#define CONTROL_THUMBS 10811
```
Would be nice to update these defines to constexpr and use an anonymous namespace but that can be done in the future.

garbear replied:
Will do in this PR

lrusak reviewed (approved):
A nice improvement!

garbear commented:
Your "Save" option is an interesting idea, if I'm right then it uses the selected savestate as a target for creating a new stavestate from the current game? Wouldn't "Overwrite" be a better term? (And more appropriately ominous given that it'll destroy the underlying save).

Instead, a family member had a good idea: Pinning savestates. They got the idea from Spotify, where you can pin up to several playlists, and the rest following are sorted chronologically. What do you think of pinning as a way to organize potentially many savestates?

KOPRajs commented:
I think that users are used to this behavior. Click "Save" on the selected savestate to overwrite it. The point is that instead of deleting the old custom labeled save, creating a new one and then renaming it again to the custom label, this "Save" feature do this in one click.

garbear commented:
So clicking "Save" on a savestate would copy it instead of overwriting it?

garbear commented:
I implemented an "Overwrite" feature which I think is your "Save" feature. Maybe "Save Here" would be a better name? Because we already have a "Save" button in the UI that does something different, it creates a new save without replacing an existing one.

garbear commented:
I think we should drop the "Save" / "Overwrite" button. I discovered that the savestate path has the date inside of it, and the new save has a newer date, so the path must change, and we can't re-use an existing path.

Secondly, if we re-use an existing path, the thumbnail's path doesn't change. This means that the cached image will continue to be shown even as a new image is written to the FS. So we really want to have the filename contain the date.

This means that a "Save" or "Overwrite" (which I think we're talking about the same thing) is just a call to "Delete" the save and then create a new save with the same label. I tested adding a "label" parameter to CreateSavestate(). Is this the approach we want to take?

We would need a new string to confirm overwriting instead of deleting, because overwriting is a destructive behavior (the original savestate is deleted). I'm not sure a "Delete-and-then-resave-with-same-label" function is worth the extra string when the user could use manually delete, and then save.

KOPRajs commented:
> So clicking "Save" on a savestate would copy it instead of overwriting it?

No, overwrite is what we want.

> Maybe "Save Here" would be a better name?

I'd say that "Save" is the usual name for this function. The difference between saving to a new slot or replacing the current one does not mean it is a different fucntion. It is the same function (it even calls the same function _CreateSavestate()_). The behavior is slightly different depending on the context, if you use the function on an empty slot (the plus button) or on an existing slot using the context menu. So IMHO the "Save" label is correct, but if you need to have a different label, then "Save here" sounds ok (please note the lowercase "h"). But if it was up to me I would re-use the existing "Save" string from translations.

> I discovered that the savestate path has the date inside of it,

I know about this. Again, I don't think it actually matters. Just think of the "Save" function as of "update the savestate" instead of "delete and create a new one". The date of the last update is changed inside the savestate while the path remains to be the "date created" instead. The original author simply used the current time for the file name instead of hash, which would probably be a better solution here. The same happens to the autosaves, the path is the time when the slot was created for the first time and is not updated when the autosave is rewritten.

I've also noticed that the picture of autosave is sometimes not updated from the same reason, but it happens only if you update the save too shortly after last update. Kodi will update the picture eventually, because the modification time of the file changes.

garbear commented:
In the future, I want to switch to a content-backed store, so changing a creation time will change the path. I think we should stick with this model, that the path changes for every save. In that case, a "Save" button will delete the existing file and create a new one at the new path. Is that what we want?

KOPRajs commented:
> In the future, I want to switch to a content-backed store, so changing a creation time will change the path. I think we should stick with this model, that the path changes for every save. In that case, a "Save" button will delete the existing file and create a new one at the new path. Is that what we want?

Do we want to change the path for manual savestates and not change it for autosaves? Why?

If I create a manual savestate as a user, then give a name to it and then use the "Save" function to update the progress manually when I feel it is a good time to save, I really don't mind that the filename is kept and that it shows the time when I created this "save slot" rather than time of the last update.

KOPRajs commented:
I think if there is anything bad about the filenames, then it is the actual use of current time for the filename. In a very rare case (when you have wrong system clock) you might accidentaly overwrite another savestate just because the current time match it's filename.

It might be better to compute a unique hash for the filename instead of using the current time.

garbear commented:
You're right, a savestate can get overwritten if the system clock changes. I didn't think about that.

Ultimately my goal is to put savestates in a content store like IPFS. You're familiar? It's an immutable content store, where the file path is a hash of the contents. Once a file is written, it's immutable forever. If a timestamp in the file changes, then it gets a new hash, and thus a new path. So in that case, the path for autosaves would change as well.

If we want paths that don't change, then we could use my pinning idea. If you pin to slot one, then you get a file like savestate_1.sav. If you unpin and then re-pin to that slot, the filename stays savestate_1.sav.

Anyways, content store is further down the road, and we should decide what to do for this feature. I think it's more helpful and cosmetic to have creation time in the filename than a hash, so clearly both auto and manual saves should overwrite the same file. We don't need to delete and create with a different name. I'll hammer out a version with our "Save" button that doesn't rename files.

I still think that it's a problem to have two "Save" buttons that do slightly different things. Does "New save" and "Save here" capture the functionality of both buttons? What would you call each button if we didn't want to use exactly "Save" twice?

garbear commented:
@KOPRajs I implemented your save button, and modified `CreateSavestate(false, savePath)` as you suggested. I don't think that "Save" is a very appropriate name for the context menu item, because saving should create a savestate, whereas here a savestate is destroyed. "Overwite" hints that the operation is destructive.

I'm working on reloading the thumbnails when the savestate changes, and unfortunately textures in Kodi are refcounted by path and can't be reloaded.  I'll have to modify the dialog and manage textures dynamically. It'll be a lot more work and code, but it's the only way to have dynamic images.

KOPRajs commented:
> "Overwrite" hints that the operation is destructive.

I still think that we are adding unnecessary work to translators because we could re-use the existing string. I believe that clicking on "Save" in context menu of the existing savestate is clear enough and user should expect that the savestate is being overwritten by this. I've even asked some other people and so far all of them suggests simply "Save". Maybe adding a confirmation dialog "Do you really want to overwrite the savestate?" after clicking the "Save" in the context menu would help?
However, if you want the second string, then both "Save (overwrite)" or "Save here" seems good.

Also have you read my comments to [75120a3](https://github.com/xbmc/xbmc/pull/22585/commits/75120a386c22eb47f11abc2654908e40ccb973fa)? I think we should drop this change and keep the possibility to rename autosaves. There is the opposite problem with renaming manual savestate to look like an autosave anyway, so this only solves half of the potential problem. Users probably won't expect that the autosave is going to stop behave like an autosave when it is renamed. Also there are legitimate reasons to rename autosaves (e. g. to distinguish players, like KOPRajs's autosave and garbear's autosave).

garbear commented:
> I still think that we are adding unnecessary work to translators because we could re-use the existing string

I actually chose overwrite because it already has a translation: https://github.com/xbmc/xbmc/pull/22585/files#diff-71c048a41de97e0c50f5e628e8dfdf8d81f40b2e9133d5824150fe3663c01d79R6558

I'm against another confirmation dialog, this leads to dialog fatigue.

I think that "Save (overwrite)" and "Save here" are better alternatives, but they would need translations - hence why I chose "Overwrite", as it's the lesser of two evils.

> Users probably won't expect that the autosave is going to stop behave like an autosave when it is renamed

This is why I made the change in the first place. When I renamed the save, it no longer said "Autosave" and there's no indication that it behaves differently than a normal save. A save should only have autosave behavior if it indicates autosave in the GUI, and when you rename the save it changes that.

garbear commented:
Here's how I plan to solve the dynamic thumbnail problem. I'm going to store the frame's pixels in the .sav file. The .jpg screenshot will simply be cosmetic for visual identification outside of Kodi. Then, I'll modify the `gamewindow` control to accept external pixels:

```xml
<control type="gamewindow">
	<width>444</width>
	<height>250</height>
	<pixels>$INFO[ListItem.Path]</pixels>
</control>
```

Because RetroPlayer writes the pixels, it knows when a pixel file is overwritten, and can update the GPU texture accordingly.

An advantage of using the `gamewindow` control instead of an `image` control is that the full suite of visual effects are applied appropriately - video filter, stretch mode and rotation.

Pixels will also greatly increase the size of the .sav file, so I'll figure out some form of compression.

KOPRajs commented:
> ... hence why I chose "Overwrite", as it's the lesser of two evils.

Ok, and I agree that too many dialogs is not good either.

> A save should only have autosave behavior if it indicates autosave in the GUI

But you can rename it to "KOPRajs's autosave" or "Autosave - KOPRajs". Actually, currently you need to intentionally delete the "Autosave" label letter by letter when renaming because you've added the _Prepopulate keyboard dialog with savestate name_.
And nothing can stop users from rename a manual savestate to "Autosave", but the Type() would be left on MANUAL in such case. So I don't think this is a good approach to change the type upon renaming.

KOPRajs commented:
> Here's how I plan to solve the dynamic thumbnail problem. I'm going to store the frame's pixels in the .sav file. The .jpg screenshot will simply be cosmetic for visual identification outside of Kodi. Then, I'll modify the `gamewindow` control to accept external pixels:

Is it worth it? Looks like a lot of work caused by a small issue. From what I've seen when using it (as autosaves are suffering from the same issue), it is not that bad and the thumbnail is updated eventually after some time and re-opening the window, isn't it?

garbear committed:
RetroPlayer: Use "screenshot" property for savestate thumbnails

garbear commented:
> But you can rename it to "KOPRajs's autosave" or "Autosave - KOPRajs". Actually, currently you need to intentionally delete the "Autosave" label letter by letter when renaming because you've added the Prepopulate keyboard dialog with savestate name. And nothing can stop users from rename a manual savestate to "Autosave", but the Type() would be left on MANUAL in such case. So I don't think this is a good approach to change the type upon renaming.

All good points. I'll drop the change but we need a way to visualize "Save type" in the GUI.

And hopefully "save type" isn't too much mental overhead, a new player might not even be familiar with what autosave is or why it's nice.

> Is it worth it?

We do these things not because they are easy, but because we thought they were going to be easy 😉

garbear commented:
What if we reverse the mental model of autosave? Where savestates by default are "auto", and we have a special icon for "bookmarked" savestates that are immutable (unless overwritten)?

garbear committed:
RetroPlayer: Rename parameter for code clarity

garbear committed:
RetroPlayer: Fix missing empty pointer check

garbear committed:
RetroPlayer: Add function to update a savestate

garbear commented:
The code to store the video frame in the savestate, and render it from the savestate, is almost done: https://github.com/garbear/xbmc/commit/35428e04c1fae7f137342d7f1e22ede6989e6928

KOPRajs commented:
> All good points. I'll drop the change but we need a way to visualize "Save type" in the GUI.

The most easy way I can think of is to split the "label" into 2 parts. The part one would be "Autosave - " or empty depending on the Type() and the second part would be the user editable label.

So we would drop adding "Autosave" label here:
https://github.com/xbmc/xbmc/blob/15a7c81e863aea2b8d4c53bf57d0fefa723a6da1/xbmc/cores/RetroPlayer/playback/ReversiblePlayback.cpp#L147

Instead we add the "Autosave" label in the presentation dialog. We can even use different color for the "Autosave" part. This should be pretty easy and the user will not be able to edit the "Autosave" part.

It seems that the following presentation is commonly used in Kodi:
![image](https://user-images.githubusercontent.com/13847813/215076239-eb8fc242-8511-49d0-986c-dab3a4caaa97.png)

So I would stick to that:
`Autosave (in the skin highlight color) dot <user label>`

Like this:
![image](https://user-images.githubusercontent.com/13847813/215079787-62681407-5ee8-421c-a7e6-2f13af095417.png)

Note that the color is not hardcoded but taken from the skin highlight color.

garbear commented:
I like the highlighted autosave. However, we really don't have room on a single line for both autosave and a custom label - think about how German text is so much longer.

Instead, let's nix the date. Date doesn't matter with autosave, because it'll always be overwritten. So the text can be a priority ladder - first label (if present), then autosave (if applicable), then date. Starting from label 1 and going to label 2 if needed, and the date will be chopped off for autosaves iff a label is present.

KOPRajs commented:
Good idea. I'm actually not sure about the highlight color, let's try it, but we may end up with the simple white in the end.

garbear committed:
RetroPlayer: Improve savestate labels

garbear commented:
Like this?

![screenshot00014](https://user-images.githubusercontent.com/531482/215174323-9457cec4-4699-48c1-90d6-3dcbffff9ffb.png)

garbear commented:
I've finished the modifications needed to achieve our vision. I think I've captured everything in this thread. I'll do builds for testing.

I only have one question. When overwriting an autosave, the save type is set to manual. Of course, if you overwrite the active autosave, then the save type is set back to auto on the next autosave. Should we preserve save type when overwriting? It would be done with the code:

```patch
--- a/xbmc/cores/RetroPlayer/playback/ReversiblePlayback.cpp
+++ b/xbmc/cores/RetroPlayer/playback/ReversiblePlayback.cpp
@@ -144,6 +144,10 @@ std::string CReversiblePlayback::CreateSavestate(bool autosave, const std::strin
       loadedSavestate.reset();
   }

+  // Update autosave flag from existing type if known
+  if (loadedSavestate)
+    autosave = (loadedSavestate->Type() == SAVE_TYPE::AUTO);
+
   const std::string caption = m_cheevos->GetRichPresenceEvaluation();
   const CDateTime nowUTC = CDateTime::GetUTCDateTime();
   const std::string gameFileName = URIUtils::GetFileName(m_gameClient->GetGamePath());
```

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

garbear commented:
I've succeeded into turning the in-game savestate thumbnails into mini-retroplayers that render the pixels stored in a savestate. It falls back to the screenshot if it can't load the pixels.

This solves the problem of savestate thumbs not refreshing (at least within the in-game OSD).

![screenshot00016](https://user-images.githubusercontent.com/531482/215236165-45660f4d-2b63-40fa-bcb4-0cb36cbe4a4c.jpg)

The color problem on 4K monitors makes it easy to identify what's a screenshot and what's being rendered by RetroPlayer. I should probably fix that...

KOPRajs commented:
Great work!

> Should we preserve save type when overwriting?

No, but when overwriting an autosave with the new "Overwrite" function, we should clear _m_autosavePath_, so the next autosave is created in a new slot and does not overwrite the just created manual savestate.

NikosSiak commented on xbmc/cores/RetroPlayer/guibridge/GUIGameSettingsHandle.h:
```diff
@@ -49,6 +49,16 @@ class CGUIGameSettingsHandle
+  /*!
+   * \brief Update a savestate for the current game being played
+   *
+   * \param path The path to the created savestate file returned by
```
minor: the param name is `savestatePath`

NikosSiak commented on xbmc/cores/RetroPlayer/guibridge/GUIGameSettingsHandle.h:
```diff
@@ -57,7 +67,7 @@ class CGUIGameSettingsHandle
    * \return True if the savestate was loaded successfully, false otherwise
    */
-  bool LoadSavestate(const std::string& path);
+  bool LoadSavestate(const std::string& savestatePath);
```
minor: change the param comment as well

NikosSiak commented on xbmc/cores/RetroPlayer/guibridge/IGameCallback.h:
```diff
@@ -36,18 +36,31 @@ class IGameCallback
+  /*!
+   * \brief Updates a savestate with the current game being played
+   *
+   * \param path The path to the savestate
```
minor: the param name is `savestatePath`

NikosSiak commented on xbmc/cores/RetroPlayer/guibridge/IGameCallback.h:
```diff
@@ -36,18 +36,31 @@ class IGameCallback
+   * \return True if the savestate was loaded, false on error
    */
-  virtual bool LoadSavestate(const std::string& path) = 0;
+  virtual bool LoadSavestate(const std::string& savestatePath) = 0;
```
minor: change the comment as well

NikosSiak reviewed (approved):
LGTM! only some minor comments about documentation

garbear commented:
@NikosSiak Updated the param names, thanks for catching those.

@KOPRajs Good suggestion, updated. In the worse case this could result in the proliferation of autosaves I mentioned earlier, but the more new files we create the better as long as we use Kodi's texture loader which can't update for overwritten images.

KOPRajs commented:
> LGTM!

How about backport? I hope we can get this to Nexus as well.

garbear commented:
Yes, both the in-game and out-of-game savestate dialogs are broken in v20. I think we should backport this for v20.1.

garbear committed:
RetroPlayer: Improve logging when memory data is missing

garbear commented:
I added a helpful error line in the log when attempting to load a savestate with no (or incorrectly sized) memory:

```
2023-01-28 17:51:42.250 T:19056   error <general>: Invalid memory size, got 823409, expected 0
```

garbear committed:
Game OSD: Improve test spacing for out-of-game savestate manager

garbear commented:
One last change, I noticed some text getting a little too close in the thumbnails, so I tweaked the spacing a bit.

Before:

![screenshot00000](https://user-images.githubusercontent.com/531482/215313646-a94616a0-7083-46b6-be99-966630407bc6.png)

After:

![screenshot00002](https://user-images.githubusercontent.com/531482/215313654-ec5888c4-3723-4b72-9fd5-712abe46ac29.png)

garbear commented:
One last change for reals, I changed the art type of savestates from icon to screenshot to avoid a conflict with savestate rendering coming next.

garbear commented:
I think we've hammered out all the issues. Is any more review wanted? Otherwise I'll merge later today.

garbear commented:
Merging, and I have another PR lined up to fix the stale screenshots in the savestate manager. If there's any issues they can be addressed in that PR.

garbear commented:
Backported: https://github.com/xbmc/xbmc/pull/22632
